### PR TITLE
[16.0][ADD] sale_product_company

### PR DESCRIPTION
**Pull request by santostelmo**
In the product template allows you to select a list of companies in which you can sell it.
Note that if the product's company field is set, it takes priority over the list of companies.

**Note**: We could add a multi-company field  `company_ids`  on product.template with the module `base_multi_company`. But that it is not an option as we don't want to have any side effect on logistics or accounting flows related product selection.

### CODE REVIEW

**Review by cyrilmanuel — APPROVED** [review]

File: sale_product_company/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    @api.model
+    def get_view(self, view_id=None, view_type="form", **options):
+        res = super().get_view(view_id, view_type, **options)
+        if view_type == "form":
+            order_xml = etree.XML(res["arch"])
+            product_id_fields = order_xml.xpath("//field[@name='product_template_id']")
+            if product_id_fields:
+                product_id_field = product_id_fields[0]
+                domain = product_id_field.get("domain", "[]")
+                domain = domain.replace(
```
Comment (simahawk) [inline sale_product_company/models/sale_order.py]: This seems quite fragile. IMO you should parse the domain, find the leaf you want to replace and rebuild the domain.

File: sale_product_company/tests/test_sale_partner_company.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+from lxml import etree
+
+from odoo.tests import Form, common
+
+
+class TestSaleProductCompany(common.TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
Comment (simahawk) [inline sale_product_company/tests/test_sale_partner_company.py:14]: tracking_disable pls

**Review by simahawk — CHANGES_REQUESTED** [review]

File: sale_product_company/models/sale_order.py (hunk 1 above)
Comment (santostelmo) [reply sale_product_company/models/sale_order.py]: Implemented something equivalent with `base_view_inheritance_extension`

**Review by simahawk — APPROVED** [review]

**Review by cyrilmanuel — APPROVED** [review]

**Comment by cyrilmanuel** [comment]
hiii @pedrobaeza  can you merge this PR please ?

**Comment by pedrobaeza** [comment]
For me, the module name is incorrect, as it doesn't reflect the usual structure of them. It should be something like `product_sellable_per_company`, starting with `product`.

**Review by jbaudoux — APPROVED** [review]

**Comment by jbaudoux** [comment]
> For me, the module name is incorrect, as it doesn't reflect the usual structure of them. It should be something like `product_sellable_per_company`, starting with `product`.

I don't agree. It depends on sale and modify the behavior of the sales order. So it should start with `sale`. I find current name appropriate

**Review by cyrilmanuel — APPROVED** [review]

**Comment by cyrilmanuel** [comment]
Hello @pedrobaeza and @jbaudoux , Could you agree on this or on the changes to be made?

**Comment by pedrobaeza** [comment]
Well, I already said about the module name, so I'm refraining from doing anything more. @simahawk can merge if you all agree it's a good module name.

**Review by gurneyalex — APPROVED** [review]

**Pull request merged**